src/Avalonia.Controls/Navigation/Frame.cs

```diff
@@ -0,0 +1,290 @@
+
+            RaiseStateProperties ( newElement );
+        }
+
+        /// <summary>
+        /// Go forward for one step.
+        /// </summary>
+        public void GoForward ()
+        {
+            if ( !CanGoForward () ) return;
+
+            var currentIndex = _history.IndexOf ( CurrentState );
+            if ( currentIndex == _history.Count () - 1 ) return;
+
+            var newElement = _history.ElementAt ( currentIndex + 1 );
+
+            ChangeContent ( newElement , CurrentState , NavigationMode.Forward );
+
+            RaiseStateProperties ( newElement );
+        }
+
+        /// <summary>
+        /// Is it possible to go back to the page?
+        /// </summary>
+        /// <returns></returns>
+        public bool CanGoBack () => _history.IndexOf ( CurrentState ) > 0;
+
+        /// <summary>
+        /// Is it possible to go to the front page?
+        /// </summary>
+        /// <returns></returns>
+        public bool CanGoForward () => _history.IndexOf ( CurrentState ) < _history.Count - 1;
+
+        /// <summary>
+        /// Current state of history.
+        /// </summary>
+        public HistoryItem CurrentState
+        {
+            get
+            {
+                return _currentState;
+            }
+            set
+            {
+                var selectedIndex = _history.IndexOf ( _currentState );
+
+                SetAndRaise ( CurrentStateProperty , ref _currentState , value );
+
+                if ( value == null ) return;
+
+                AddNewItemToHistory ( value , selectedIndex );
+
+                ChangeContent ( value , null , NavigationMode.New );
+
+                _currentView = value.Type;
+                RaisePropertyChanged ( CurrentViewProperty , CurrentView , value.Type );
+            }
+        }
+
+        /// <summary>
+        /// Current view of history (if you need a simple shift page, without parameters).
+        /// </summary>
+        public Type CurrentView
+        {
+            get
+            {
+                return _currentView;
+            }
+            set
+            {
+                var selectedIndex = _history.IndexOf ( _currentState );
+
+                SetAndRaise ( CurrentViewProperty , ref _currentView , value );
+
+                if ( value == null ) return;
+
+                var historyItem = new HistoryItem {
+                    Type = value ,
+                    Parameters = Enumerable.Empty<HistoryItem> ()
+                };
+
+                AddNewItemToHistory ( historyItem , selectedIndex );
+
+                ChangeContent ( historyItem , null , NavigationMode.New );
+
+                _currentState = historyItem;
+                RaisePropertyChanged ( CurrentStateProperty , CurrentState , historyItem );
+            }
+        }
+
+        /// <summary>
+        /// This parameter influence on 
+        /// </summary>
+        public bool IsStack
```

Comment:
Why dont you introduce a cache for used pages and let us set the cache size. By default the size could be one.